Feature project report detail

maciejSamerdak opened the pull request:
Introduces report editing view linked to **project-report-list**.

Related #133

maciejSamerdak commented:
Due to lack of admin's serializer, I changed this view to a Django-based generic view, in resemblance to AdminReportView.

rwrzesien commented on employees/tests/test_unit_report_viewset.py:
```diff
@@ -642,3 +643,100 @@ def test_project_report_list_get_queryset_method_should_return_queryset_containi
+        self.assertContains(response, self.report.description)
+        self.assertEqual(response.data["serializer"].instance, self.report)
+
+    def test_project_report_detail_view_should_not_be_accessible_for_unauthenticated_users(self):
```
I think this test is not needed, `login_decorator` is working well :) Lets just test our custom logic.

maciejSamerdak replied:
You think so? This way at least we know that the view isn't protected anymore if anything goes wrong. Or are there other, better ways to test that?

rwrzesien replied:
Right, lets keep it then.

rwrzesien commented on employees/tests/test_unit_report_viewset.py:
```diff
@@ -642,3 +643,100 @@ def test_project_report_list_get_queryset_method_should_return_queryset_containi
+        response = ProjectReportDetail.as_view()(request, pk=self.report.pk)
+        self.assertEqual(response.status_code, 403)
+
+    def test_project_report_detail_view_should_not_render_non_existing_report(self):
```
Same here, routing should work OK by default too, I think this test is not necessary.

maciejSamerdak replied:
Done

rwrzesien commented on employees/tests/test_unit_report_viewset.py:
```diff
@@ -642,3 +643,100 @@ def test_project_report_list_get_queryset_method_should_return_queryset_containi
+
+class ProjectReportDetailTests(TestCase):
+    def setUp(self):
+        self.user = CustomUser(
```
Please use `UserFactory`.

maciejSamerdak replied:
Done

rwrzesien commented on employees/tests/test_unit_report_viewset.py:
```diff
@@ -642,3 +643,100 @@ def test_project_report_list_get_queryset_method_should_return_queryset_containi
+        self.project.full_clean()
+        self.project.save()
+
+        self.report = Report(
```
Please user `ReportFactory`. Then you don't have to create `self.project` too.

maciejSamerdak replied:
Done

rwrzesien commented on employees/templates/employees/project_report_detail.html:
```diff
@@ -0,0 +1,26 @@
+        {% csrf_token %}
+        {{ form|crispy }}
+        <input type="submit" value="{{ UI_text.UPDATE_REPORT_BUTTON.value }}">
+        <input type="submit" name="discard" value="{{ UI_text.DISCARD_CHANGES_BUTTON.value }}">
```
Is this being handled by Django by default? 

Generally I don't like making a POST request just to **not** make any changes. I think from UX perspective it is simpler to "go back" or "cancel" or something like that. I would suggest replacing it with a link back.

maciejSamerdak replied:
I'm happy to inform you that everything is in absolute order and this solution in fact does not work at all. 😄 

Not only does the test fail, I also cannot see any other button except for **Update** in rendered view. I guess I must have skipped the unit tests on this branch, during the hassle of multiple rebases.

maciejSamerdak replied:
Well, I'm happy to inform you that it does not work at all. 😁 

The _discard_ button just works as another submit, changes apply. I'll fix it up in a jiff.

maciejSamerdak replied:
Done

rwrzesien commented on employees/tests/test_unit_report_viewset.py:
```diff
@@ -642,3 +643,100 @@ def test_project_report_list_get_queryset_method_should_return_queryset_containi
+        self.report.save()
+
+    def test_project_report_detail_view_should_display_report_details(self):
+        request = APIRequestFactory().get(path=reverse("project-report-detail", args=(self.report.pk,)))
```
Please use `self.client` that is delivered with Django `TestCase`. You can simply make `self.client.force_login(self.user)` to make it authenticated (ie. in `setUp()`). You can take a look how it is used in other refactored tests for Django views.

maciejSamerdak replied:
Done

rwrzesien commented on employees/tests/test_unit_report_viewset.py:
```diff
@@ -642,3 +643,100 @@ def test_project_report_list_get_queryset_method_should_return_queryset_containi
+        request = APIRequestFactory().post(
+            path=reverse("project-report-detail", args=(self.report.pk,)),
+            data={
+                "date": datetime.datetime.now().date(),
```
If its reused (or part of it is), add it to `setUp()` like `self.data`.

maciejSamerdak replied:
Done

rwrzesien commented:
@maciejSamerdak Looks nice, please just adapt the test and please explain me how the "discard" works :)

maciejSamerdak commented:
@rwrzesien I added requested changes. I also refactored unit tests a little and moved them to `test_views`. However, I'd like to highlight one issue: after refactoring the tests, they performance dropped significantly. It's like a double-action revolver vs a Tommy gun.

rwrzesien commented:
@maciejSamerdak 
> It's like a double-action revolver vs a Tommy gun.

Nice comparison :) Well that is because previously we was running view directly as a function. Now it is sending request through all other stuff like middleware, routing etc. - basically in the same way as you would send a request to Django server running. It takes slower **but** also behaves more like a regular HTTP call, allowing us to test/discover more stuff (for example if any other layer has impact on the view).

rwrzesien commented:
@maciejSamerdak Looks good, tests are much more compact now, this is exactly what I have meant. If you want add back the test `test_project_report_detail_view_should_not_be_accessible_for_unauthenticated_users` but if you don't want its fine for me too.

maciejSamerdak commented:
I recovered the login-required test, just in case. When integration tests are introduced, covering this mechanism in greater detail, it could be removed.

kbeker commented:
Accidentally closed. Merged
